kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -69,10 +71,15 @@ public class TransactionImpl implements Transaction {
   private final Optional<SetTransaction> setTxnOpt;
   private final boolean shouldUpdateProtocol;
   private final Clock clock;
-  private final List<DomainMetadata> domainMetadatas = new ArrayList<>();
   private Metadata metadata;
   private boolean shouldUpdateMetadata;
 
+  // Contains domain metadata actions known prior to iterating and writing the data actions
+  private final List<DomainMetadata> domainMetadatas = new ArrayList<>();
+  // Contains domain metadata actions generated on the fly while writing the data actions
+  private CloseableIterator<DomainMetadata> domainMetadataIter =
+      toCloseableIterator(Collections.emptyIterator());
```

Comment:
I probably will change this. Currently the list is only used in tests, and iterator is only used for row tracking's domain metadata. I’ll look for more use cases of domain metadata in Delta-Spark to see if there is a better way for managing `DomainMetadata` actions in a transaction.